File: src/scanner.rs
use std::collections::HashMap;

use crate::{ literal::Literal, everest::Everest, token::{ Token, TokenKind } };

pub struct Scanner {
    source: String,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
    lox: Everest,
}

impl Scanner {
    pub fn new(source: String, lox: Everest) -> Self {
        Self {
            source,
            tokens: vec![],
            current: 0,
            line: 0,
            start: 0,
            lox,
        }
    }

    pub fn scan_tokens(&mut self) -> &Vec<Token> {
        // println!("{}", self.source);
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token();
        }

        self.tokens.push(Token::new(TokenKind::Eof, "".to_string(), Literal::None, self.line));

        &self.tokens
    }

    fn is_at_end(&mut self) -> bool {
        self.current >= self.source.len()
    }

    pub fn scan_token(&mut self) {
        let c: char = self.advance();

        match c {
            '(' => self.put_token(TokenKind::LeftParen),
            ')' => self.put_token(TokenKind::RightParen),
            '{' => self.put_token(TokenKind::LeftBrace),
            '}' => self.put_token(TokenKind::RightBrace),
            ',' => self.put_token(TokenKind::Comma),
            '.' => self.put_token(TokenKind::Dot),
            '-' => self.put_token(TokenKind::Minus),
            '+' => self.put_token(TokenKind::Plus),
            ';' => self.put_token(TokenKind::Semicolon),
            '*' => self.put_token(TokenKind::Star),
            '!' => {
                let tok = self
                    .is_next('=')
                    .then(|| TokenKind::Ne)
                    .unwrap_or(TokenKind::Bang);
                self.put_token(tok);
            }
            '=' => {
                let tok = self
                    .is_next('=')
                    .then(|| TokenKind::EqEq)
                    .unwrap_or(TokenKind::Eq);
                self.put_token(tok);
            }
            '<' => {
                let tok = self
                    .is_next('=')
                    .then(|| TokenKind::Le)
                    .unwrap_or(TokenKind::Lt);
                self.put_token(tok);
            }
            '>' => {
                let tok = self
                    .is_next('=')
                    .then(|| TokenKind::Ge)
                    .unwrap_or(TokenKind::Gt);
                self.put_token(tok);
            }
            '/' => {
                if self.is_next('/') {
                    while self.peek() != '\n' && !self.is_at_end() {
                        self.advance();
                    }
                } else {
                    self.put_token(TokenKind::Slash)
                }
            }
            ' ' | '\r' | '\t' => {}
            '\n' => {
                self.line += 1;
            }

            '"' => self.string(),

            c if c.is_ascii_digit() => {
                self.number();
            }

            c if c.is_alphabetic() => self.identifier(),

            c => self.lox.error(self.line, format!("Unexpected character '{c}'")),
        }
    }

    fn advance(&mut self) -> char {
        let i = {
            let temp = self.current;
            self.current += 1;
            temp
        };

        self.source[i..i + 1].chars().next().unwrap()
    }

    fn add_token(&mut self, kind: TokenKind, literal: Literal) {
        let text = &self.source[self.start..self.current];

        self.tokens.push(Token::new(kind, text.to_string(), literal, self.line))
    }

    fn put_token(&mut self, kind: TokenKind) {
        self.add_token(kind, Literal::None);
    }

    fn is_next(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }

        if self.peek() != expected {
            return false;
        }

        self.current += 1;
        return true;
    }

    pub fn peek(&mut self) -> char {
        if self.is_at_end() {
            return '\0';
        }
        return self.source.chars().nth(self.current).unwrap();
    }

    fn string(&mut self) {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }

            self.advance();
        }

        if self.is_at_end() {
            self.lox.error(self.line, format!("Unterminated string"));
            return;
        }

        self.advance();

        let value = &self.source[self.start + 1..self.current - 1];
        self.add_token(TokenKind::String, Literal::String(value.to_string()));
    }

    fn number(&mut self) {
        while self.peek().is_ascii_digit() {
            self.advance();
        }

        if self.peek() == '.' && self.peek_next().is_ascii_digit() {
            self.advance();

            while self.peek().is_ascii_digit() {
                self.advance();
            }
        }

        let mov = &self.source[self.start..self.current];

        self.add_token(TokenKind::Number, Literal::Number(mov.parse().unwrap()))
    }

    pub fn peek_next(&mut self) -> char {
        self.source
            .chars()
            .nth(self.current + 1)
            .unwrap_or('\0')
    }

    pub fn identifier(&mut self) {
        while self.peek().is_ascii_alphanumeric() {
            self.advance();
        }

        let text = &self.source[self.start..self.current];
        let h = Self::keywords();
        let kind = h.get(text).unwrap_or(&TokenKind::Identifier);

        self.put_token(*kind)
    }

    pub fn keywords() -> HashMap<String, TokenKind> {
        let mut hash = HashMap::new();

        hash.insert("and", TokenKind::And);
        hash.insert("class", TokenKind::Class);
        hash.insert("else", TokenKind::Else);
        hash.insert("false", TokenKind::False);
        hash.insert("for", TokenKind::For);
        hash.insert("fun", TokenKind::Fn);
        hash.insert("if", TokenKind::If);
        hash.insert("nil", TokenKind::None);
        hash.insert("or", TokenKind::Or);
        hash.insert("print", TokenKind::Print);
        hash.insert("return", TokenKind::Return);
        hash.insert("super", TokenKind::Super);
        hash.insert("this", TokenKind::This);
        hash.insert("true", TokenKind::True);
        hash.insert("var", TokenKind::Var);
        hash.insert("while", TokenKind::While);

        hash.into_iter()
            .map(|(k, v)| (k.to_string(), v))
            .collect()
    }
}

File: js/src/old/scanner.ts
import { match } from '@rqft/rust';
import { Everest } from './everest';
import type { Literal } from './token';
import { Token } from './token';
import { TokenKind } from './token_kind';
export class Scanner {
  private readonly source: string;
  private readonly tokens: Array<Token> = [];
  private start = 0;
  private current = 0;
  private line = 1;

  constructor(source: string) {
    this.source = source;
  }

  public scan_tokens(): Array<Token> {
    while (!this.is_at_end()) {
      this.start = this.current;
      this.scan_token();
    }

    this.tokens.push(new Token(TokenKind.Eof, '', null, this.line));
    return this.tokens;
  }

  private is_at_end(): boolean {
    return this.current >= this.source.length;
  }

  private scan_token(): void {
    const c: string = this.next();
    match<string, void>(c)
      .arm('(', () => this.add_token(TokenKind.LeftParen))
      .arm(')', () => this.add_token(TokenKind.RightParen))
      .arm('{', () => this.add_token(TokenKind.LeftBrace))
      .arm('}', () => this.add_token(TokenKind.RightBrace))
      .arm(',', () => this.add_token(TokenKind.Comma))
      .arm('.', () => this.add_token(TokenKind.Dot))
      .arm('-', () => this.add_token(TokenKind.Minus))
      .arm('+', () => this.add_token(TokenKind.Plus))
      .arm(';', () => this.add_token(TokenKind.Semi))
      .arm('*', () => this.add_token(TokenKind.Star))
      .arm('!', () =>
        this.add_token(this.is('=') ? TokenKind.Ne : TokenKind.Bang)
      )
      .arm('=', () =>
        this.add_token(this.is('=') ? TokenKind.EqEq : TokenKind.Eq)
      )
      .arm('<', () =>
        this.add_token(this.is('=') ? TokenKind.Le : TokenKind.Lt)
      )
      .arm('>', () =>
        this.add_token(this.is('=') ? TokenKind.Ge : TokenKind.Gt)
      )
      .arm('/', () => {
        if (this.is('/')) {
          while (this.peek() !== '\n' && !this.is_at_end()) {
            this.next();
          }
        } else {
          this.add_token(TokenKind.Slash);
        }
      })
      .union([' ', '\r', '\t'], undefined)
      .arm('\n', () => this.line++)
      .arm('"', () => this.string())

      .wildcard((c) => {
        if (this.is_digit(c)) {
          this.number();
        } else if (this.is_alpha(c)) {
          this.identifier();
        } else {
          Everest.error(this.line, 'unexpected char');
        }
      })
      .output();
  }

  private identifier(): void {
    while (this.is_alphanumeric(this.peek())) {
      this.next();
    }

    const text = this.source.substring(this.start, this.current);
    const type = this.keywords().get(text) || TokenKind.Identifier;

    this.add_token(type);
  }

  private keywords(): Map<string, TokenKind> {
    const map = new Map<string, TokenKind>();

    map.set('and', TokenKind.And);
    map.set('class', TokenKind.Class);
    map.set('else', TokenKind.Else);
    map.set('false', TokenKind.False);
    map.set('for', TokenKind.For);
    map.set('fn', TokenKind.Fn);
    map.set('if', TokenKind.If);
    map.set('none', TokenKind.None);
    map.set('or', TokenKind.Or);
    map.set('print', TokenKind.Print);
    map.set('return', TokenKind.Return);
    map.set('super', TokenKind.Super);
    map.set('this', TokenKind.This);
    map.set('true', TokenKind.True);
    map.set('var', TokenKind.Var);
    map.set('while', TokenKind.While);

    return map;
  }

  private is_alpha(c: string): boolean {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
  }

  private is_alphanumeric(c: string): boolean {
    return this.is_digit(c) || this.is_alpha(c);
  }

  private is_digit(char: string): boolean {
    return char >= '0' && char <= '9';
  }

  private number(): void {
    while (this.is_digit(this.peek())) {
      this.next();
    }

    if (this.peek() === '.' && this.is_digit(this.peek_next())) {
      this.next();

      while (this.is_digit(this.peek())) {
        this.next();
      }
    }

    this.add_token_with(
      TokenKind.Number,
      Number(this.source.substring(this.start, this.current))
    );
  }

  private peek_next(): string {
    if (this.current + 1 >= this.source.length) {
      return '\0';
    }

    return this.source.charAt(this.current + 1);
  }

  private string(): void {
    while (this.peek() !== '"' && !this.is_at_end()) {
      if (this.peek() === '\n') {
        this.line++;
      }
      this.next();
    }

    if (this.is_at_end()) {
      Everest.error(this.line, 'unterminated string');
      return;
    }

    this.next();

    const value = this.source.substring(this.start + 1, this.current - 1);
    this.add_token_with(TokenKind.String, value);
  }

  private next(): string {
    return this.source.charAt(this.current++);
  }

  private add_token(kind: TokenKind): void {
    this.add_token_with(kind, null);
  }

  private add_token_with(kind: TokenKind, literal: Literal): void {
    const text = this.source.substring(this.start, this.current);
    this.tokens.push(new Token(kind, text, literal, this.line));
  }

  private is(expected: string): boolean {
    if (this.is_at_end()) {
      return false;
    }

    if (this.source.charAt(this.current) !== expected) {
      return false;
    }

    this.current++;
    return true;
  }

  private peek(): string {
    if (this.is_at_end()) {
      return '\0';
    }

    return this.source.charAt(this.current);
  }
}
